bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1387 @@
+						MAX_SZ, UNDEF_MULTPL},
+	[BGP_LS_TLV_LINK_LOCAL_REMOTE_IDENTIFIERS] =
+		{"Link Local/Remote Identifiers", 2, 2, UNDEF_MULTPL},
+	[BGP_LS_TLV_IPV4_INTERFACE_ADDRESS] = {"IPv4 interface address", 4, 4,
+					       UNDEF_MULTPL},
+	[BGP_LS_TLV_IPV4_NEIGHBOR_ADDRESS] = {"IPv4 neighbor address", 4, 4,
+					      UNDEF_MULTPL},
+	[BGP_LS_TLV_IPV6_INTERFACE_ADDRESS] = {"IPv6 interface address", 16, 16,
+					       UNDEF_MULTPL},
+	[BGP_LS_TLV_IPV6_NEIGHBOR_ADDRESS] = {"IPv6 neighbor address", 16, 16,
+					      UNDEF_MULTPL},
+	[BGP_LS_TLV_OSPF_ROUTE_TYPE] = {"OSPF Route Type", 1, 1, UNDEF_MULTPL},
+	[BGP_LS_TLV_IP_REACHABILITY_INFORMATION] =
+		{"IP Reachability Information", 2, 17, UNDEF_MULTPL},
+	[BGP_LS_TLV_AUTONOMOUS_SYSTEM] = {"Autonomous System", 4, 4,
+					  UNDEF_MULTPL},
+	[BGP_LS_TLV_BGP_LS_IDENTIFIER] = {"BGP-LS Identifier", 4, 4,
+					  UNDEF_MULTPL},
+	[BGP_LS_TLV_OSPF_AREA_ID] = {"OSPF Area-ID", 4, 4, UNDEF_MULTPL},
+	[BGP_LS_TLV_IGP_ROUTER_ID] = {"IGP Router-ID", 4, 8, UNDEF_MULTPL},
+	/* NRLI & BGP-LS Attributes */
+	[BGP_LS_TLV_MULTI_TOPOLOGY_ID] = {"Multi-Topology ID", 2, MAX_SZ, 2},
+};
+
+/* Return the TLV length is valid for the TLV type */
+static bool bgp_ls_tlv_check_size(enum bgp_linkstate_tlv type, size_t length)
+{
+	if (type > BGP_LS_TLV_MAX ||
+	    bgp_linkstate_tlv_infos[type].descr == NULL)
+		/* TLV type is not defined. Cannot check size */
+		return false;
+
+	if (bgp_linkstate_tlv_infos[type].min_size > length)
+		return false;
+	if (bgp_linkstate_tlv_infos[type].max_size < length)
+		return false;
+	if (length % bgp_linkstate_tlv_infos[type].multiple != 0)
+		return false;
+
+	return true;
+}
+
+static uint8_t pnt_decode8(uint8_t **pnt)
+{
+	uint8_t data;
+
+	data = **pnt;
+	*pnt += 1;
+	return data;
+}
+
+static uint16_t pnt_decode16(uint8_t **pnt)
+{
+	uint16_t data;
+
+	*pnt = ptr_get_be16(*pnt, &data);
+
+	return data;
+}
+
+static uint32_t pnt_decode32(uint8_t **pnt)
+{
+	uint32_t data;
+
+	*pnt = (uint8_t *)ptr_get_be32(*pnt, &data);
+
+	return data;
+}
+
+static uint64_t pnt_decode64(uint8_t **pnt)
+{
+	uint64_t data;
+
+	*pnt = (uint8_t *)ptr_get_be64(*pnt, &data);
+
+	return data;
+}
+
+static const char *bgp_ls_print_nlri_proto(enum bgp_ls_nlri_proto proto)
+{
+	switch (proto) {
+	case BGP_LS_NLRI_PROTO_ID_IS_IS_LEVEL_1:
+		return "ISIS-L1";
+	case BGP_LS_NLRI_PROTO_ID_IS_IS_LEVEL_2:
+		return "ISIS-L2";
+	case BGP_LS_NLRI_PROTO_ID_OSPF:
+		return "OSPFv2";
+	case BGP_LS_NLRI_PROTO_ID_DIRECT:
+		return "Direct";
+	case BGP_LS_NLRI_PROTO_ID_STATIC:
+		return "Static";
+	case BGP_LS_NLRI_PROTO_ID_OSPFv3:
+		return "OSPFv3";
+	case BGP_LS_NLRI_PROTO_ID_UNKNOWN:
+		return "Unknown";
+	}
+	return "Unknown";
+}
+
+static int bgp_linkstate_nlri_node_descriptor_decode(
+	struct bgp_ls_nlri_node_descr_tlv *node_descr, uint8_t *pnt,
+	uint16_t len)
+{
+	uint8_t *lim = pnt + len;
+	uint16_t type, length;
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
```

Comment:
It would be `pnt = (uint8_t *)ptr_get_be16(pnt, &type);`

I am incrementing the `pnt` pointer by two bytes after writing to `type`

I have code for a future pull request that displays data from the stream like that

`json_object_int_add(json_data, "rangeSize", pnt_decode16(&pnt));`

`pnt_decodeX()` functions are more convenient